### PR TITLE
Improve smart defaults for .NET Core apps

### PR DESCRIPTION
**Pull request by nturinski**
Fixes https://github.com/microsoft/vscode-azureappservice/issues/945

Right now, I'm just throwing the bits into a folder called `publish`.  But I'm contemplating whether I should put the output in a temp folder or clean up the bits after deployment.

The default will be defined by the `tasks.json`.  If the `preDeployTask` is already defined, then we assume that the task exists.  If it doesn't, this error comes up:

image

### CODE REVIEW

**Comment by nturinski** [comment]
Offline feedback:
- Publish in a temp folder-- clean post-deployment
- Default web app to Linux (in the original issue)
- If the user deploys the published bits themselves, don't set preDeployTask for them

File: src/commands/createWebApp/setAppWizardContextDefault.ts [hunk 1]
```diff
@@ -35,6 +36,9 @@ export async function setAppWizardContextDefault(wizardContext: IAppServiceWizar
             // requirements.txt are used to pip install so a good way to determine it's a Python app
             wizardContext.recommendedSiteRuntime = [LinuxRuntimes.python];
 
+        } else if ((await findFilesByFileExtension(fsPath, 'csproj')).length > 0) {
+            // wizardContext.recommendedSiteRuntime = ['dotnet'];
```
Comment (nturinski) [inline src/commands/createWebApp/setAppWizardContextDefault.ts]: Will need to uncomment this when this PR is in https://github.com/microsoft/vscode-azuretools/pull/528

File: src/commands/deploy/setPreDeployTaskForDotnet.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    const dotnetOutputPath: string = path.join('bin', 'Debug', 'publish');
+
+    // tslint:disable-next-line: strict-boolean-expressions
+    if (!currentWorkspace || !getWorkspaceSetting('configurePreDeployTasks', currentWorkspace.uri.fsPath)) {
+        // if the workspace being deployed is not opened, return and do nothing
+        return;
+    }
+
+    const csProj = await workspaceUtil.findFilesByFileExtension(currentWorkspace.uri.fsPath, 'csproj');
+    if (csProj.length > 0) {
```
Comment (nturinski) [inline src/commands/deploy/setPreDeployTaskForDotnet.ts]: Would it be better to check the siteConfig.linuxFxVersion is .NET? Or both?

**Comment by nturinski** [comment]
Offline feedback:
Don't overwrite the preDeployTask if they already have something there.

File: src/commands/deploy/setPreDeployTaskForDotnet.ts (hunk 2 above)
Comment (EricJizbaMSFT) [reply src/commands/deploy/setPreDeployTaskForDotnet.ts]: Good question - I'd have to think about it more. Can you explain why you chose what you did?

File: src/commands/deploy/setPreDeployTaskForDotnet.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    if (!getWorkspaceSetting('configurePreDeployTasks', currentWorkspace.uri.fsPath)) {
+        return;
+    }
+
+    const csProj = await workspaceUtil.findFilesByFileExtension(currentWorkspace.uri.fsPath, 'csproj');
+    if (csProj.length > 0) {
+        // if a publish task is already defined, then assume that we don't need this logic
+        if (getWorkspaceSetting(constants.configurationSettings.preDeployTask, currentWorkspace.uri.fsPath) !== 'publish') {
+            // if this doesn't have the publish preDeployTask, configure for .NET depoyment
+            const csProjDoc: TextDocument = await workspace.openTextDocument(csProj[0]);
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/setPreDeployTaskForDotnet.ts]: I usually prefer `fs-extra` for reading a file instead of VS Code apis

File: src/commands/deploy/setPreDeployTaskForDotnet.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            const csProjJson: string = csProjDoc.getText();
+
+            const tfw: string = 'TargetFrameworks';
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/setPreDeployTaskForDotnet.ts]: I'm not sold on this logic. I think it would be confusing to prompt for a target framework out of the blue and this feels like an edge case.

I would rather handle the more standard case like this: https://github.com/Azure-Samples/dotnet-core-api/blob/master/TodoApi.csproj#L4 and wait for feedback/telemetry before we get into more complex cases.

File: src/commands/deploy/setPreDeployTaskForDotnet.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    const fsPath: string = nonNullProp(context, 'fsPath');
+    const currentWorkspace: WorkspaceFolder | undefined = nonNullValue(workspaceUtil.getContainingWorkspace(fsPath));
+    // follow the publish output patterns, but leave out tfw
+    const dotnetOutputPath: string = path.join('bin', 'Debug', 'publish');
+
+    if (!getWorkspaceSetting('configurePreDeployTasks', currentWorkspace.uri.fsPath)) {
+        return;
+    }
+
+    const csProj = await workspaceUtil.findFilesByFileExtension(currentWorkspace.uri.fsPath, 'csproj');
```
[the remark this answers is not in the document]
Comment (nturinski) [reply src/commands/deploy/setPreDeployTaskForDotnet.ts]: I believe that this only looks in the root folder.  I'm using the vscode.workspace.findFiles with a relative pattern.

File: src/commands/deploy/setPreDeployTaskForDotnet.ts (hunk 2 above)
Comment (nturinski) [reply src/commands/deploy/setPreDeployTaskForDotnet.ts]: I think the .csproj file is what's use to detect whether or not something is a C# project in the Omnisharp extension @fiveisprime at least recommended that's what we do.  

A publish will fail if we don't have the .csproj file, so I guess it makes sense to check for that.

File: src/commands/deploy/setPreDeployTaskForDotnet.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply src/commands/deploy/setPreDeployTaskForDotnet.ts]: Well I think it's pretty common to put it in a sub folder so this is worth figuring out

File: src/commands/createWebApp/setAppWizardContextDefault.ts (hunk 1 above)
Comment (EricJizbaMSFT) [reply src/commands/createWebApp/setAppWizardContextDefault.ts]: That was merged a while ago - anything you're waiting on to fix this comment?

File: src/commands/deploy/setPreDeployTaskForDotnet.ts (hunk 2 above)
Comment (EricJizbaMSFT) [reply src/commands/deploy/setPreDeployTaskForDotnet.ts]: Omnisharp is a completely different scenario IMO. They _add_ .NET functionality to your workspace, but don't block other languages. Here we have to choose one and _only_ one language used for deploying.

We already have logic to detect if a project is a node.js app (aka if it has a package.json) or java app. I think you need to somehow refactor so that if a workspace matches multiple languages (aka both a csproj and a package.json) you can bail out and not use the defaults for either language.

As an alternative, perhaps you _should_ check `siteConfig.linuxFxVersion` since that seems more deterministic. But that would only work for linux and might have other drawbacks

File: src/commands/deploy/deploy.ts [hunk 6]
```diff
@@ -7,53 +7,40 @@ import * as WebSiteModels from 'azure-arm-website/lib/models';
...
-        siteConfig = node ? await node.root.client.getSiteConfig() : undefined;
-
-        if (siteConfig && javaUtils.isJavaRuntime(siteConfig.linuxFxVersion)) {
-            const fileExtension: string = javaUtils.getArtifactTypeByJavaRuntime(siteConfig.linuxFxVersion);
-            context.fsPath = await workspaceUtil.showWorkspaceFolders(`Select the ${fileExtension} file to deploy...`, context, constants.configurationSettings.deploySubpath, fileExtension);
-        } else {
-            context.fsPath = await workspaceUtil.showWorkspaceFolders("Select the folder to deploy", context, constants.configurationSettings.deploySubpath);
-        }
-    }
+    context.workspace = currentWorkspace;
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/deploy.ts]: How were you able to get rid of the java logic?

File: src/commands/createWebApp/setAppWizardContextDefault.ts [hunk 7]
```diff
@@ -5,28 +5,30 @@
...
 
 export interface IDeployWizardContext extends IActionContext {
-    fsPath?: string;
+    workspace: WorkspaceFolder;
     deployedWithConfigs?: boolean;
     configurationTarget?: ConfigurationTarget;
 }
 
-export async function setAppWizardContextDefault(wizardContext: IAppServiceWizardContext & IDeployWizardContext): Promise<void> {
+export async function setAppWizardContextDefault(wizardContext: IAppServiceWizardContext, deployedWorkspace?: WorkspaceFolder): Promise<void> {
```
Comment (EricJizbaMSFT) [inline src/commands/createWebApp/setAppWizardContextDefault.ts]: `wizardContext: IAppServiceWizardContext & Partial<IDeployWizardContext>`

and then you should be able to use `wizardContext.workspace`

File: src/commands/deploy/setPreDeployTaskForDotnet.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import { openUrl } from '../../utils/openUrl';
+import * as workspaceUtil from '../../utils/workspace';
+import { getWorkspaceSetting, updateWorkspaceSetting } from '../../vsCodeConfig/settings';
+import * as tasks from '../../vsCodeConfig/tasks';
+import { IDeployWizardContext } from "../createWebApp/setAppWizardContextDefault";
+
+export async function setPreDeployTaskForDotnet(context: IDeployWizardContext): Promise<void> {
+    await validateDotnetInstalled(context);
+    // follow the publish output patterns, but leave out tfw
+    const dotnetOutputPath: string = path.join('bin', 'Debug', 'publish');
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/setPreDeployTaskForDotnet.ts]: How were you able to get rid of all tfw logic?

Comment (nturinski) [reply src/commands/deploy/setPreDeployTaskForDotnet.ts]: If we just rely on `dotnet publish`, we don't need to specify the framework. The output path is something that we are making ourselves.

In a multiple target framework scenario, the publish will fail telling the user to specify their twf

**Comment by nturinski** [comment]
Offline feedback:
Use the web app's siteConfig if we are deploying to determine the site's type.  It should be more reliable than parsing the project.

Look for the .csproj file, but if we can't find it, do the following:
- In Linux case: Request the csproj OR the publish folder
- In Windows case: Don't set up any defaults

Do NOT block users from deploying .NET projects if they don't have .NET core installed.  There's no guarantee that our method of finding .NET is 100% accurate.  Plus they might just have the published bits received from somebody else.

File: src/commands/deploy/setPreDeployTaskForDotnet.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+    // if there was no .csproj in the root space, try to find it within subfolders
+    if (csProjInRoot.length === 0) {
+        // to have a recursive search of the opened workspace, pass in undefined rather than the fsPath
+        csProjInSubfolders = await findFilesByFileExtension(undefined, 'csproj');
+        try {
+            if (csProjInSubfolders.length === 1) {
+                csProjFsPath = csProjInSubfolders[0].fsPath;
+            } else if (csProjInSubfolders.length > 1) {
+                csProjFsPath = (await ext.ui.showQuickPick(mapFilesToQuickPickItems(csProjInSubfolders), { placeHolder: 'Select a .csproj file' })).data;
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/setPreDeployTaskForDotnet.ts]: It seems out-of-the-blue and not obvious why I'm selecting a csproj file. I would prefer not to prompt for now and just assume you didn't find one. Just add telemetry so that we know how often this happens and if we need to handle it.

File: src/commands/deploy/setPreDeployTaskForDotnet.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    // assume that the csProj is in the root
+    let csProjFsPath: string = context.workspace.uri.fsPath;
+
+    const csProjInRoot: Uri[] = await findFilesByFileExtension(context.workspace.uri.fsPath, 'csproj');
+    let csProjInSubfolders: Uri[] = [];
+
+    // if there was no .csproj in the root space, try to find it within subfolders
+    if (csProjInRoot.length === 0) {
+        // to have a recursive search of the opened workspace, pass in undefined rather than the fsPath
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/setPreDeployTaskForDotnet.ts]: This is completely recursive? That could really slow down deployment if you have a big workspace and I don't think we want to do that.

File: src/commands/createWebApp/setDefaultRgAndPlanName.ts [hunk 11]
```diff
@@ -91,12 +92,12 @@ async function promptPerformanceWarning(context: IActionContext, asp: AppService
...
 
         if (input === AppServiceDialogResponses.turnOnAdvancedCreation) {
-            await workspaceConfig.update('advancedCreation', true, ConfigurationTarget.Global);
+            await updateGlobalSetting(configurationSettings.advancedCreation, true);
             context.telemetry.properties.cancelStep = AppServiceDialogResponses.turnOnAdvancedCreation.title;
             throw new UserCancelledError();
         } else if (input === DialogResponses.dontWarnAgain) {
             context.telemetry.properties.turnOffPerfWarning = 'true';
-            workspaceConfig.update(showPlanPerformanceWarningSetting, false, ConfigurationTarget.Global);
+            await updateGlobalSetting(showPlanPerformanceWarningSetting, false);
```
Comment (EricJizbaMSFT) [inline src/commands/createWebApp/setDefaultRgAndPlanName.ts]: Any chance you could move these changes to a separate PR? The addition of "../../vsCodeConfig/settings"; affects a lot of files so I find myself re-reviewing it over and over, but it's pretty simple stuff by itself and could've been merged already

File: src/commands/deploy/getDeployFsPath.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import * as vscode from 'vscode';
+import { DialogResponses, IActionContext } from 'vscode-azureextensionui';
+import { configurationSettings, extensionPrefix } from '../../constants';
+import { SiteTreeItem } from '../../explorer/SiteTreeItem';
+import { ext } from '../../extensionVariables';
+import { isPathEqual, isSubpath } from '../../utils/pathUtils';
+import * as workspaceUtil from '../../utils/workspace';
+import { getWorkspaceSetting, updateGlobalSetting } from '../../vsCodeConfig/settings';
+
+export async function getDeployFsPath(context: IActionContext, target: vscode.Uri | string | SiteTreeItem | undefined, fileExtension?: string): Promise<string> {
```
Comment (EricJizbaMSFT) [inline src/commands/deploy/getDeployFsPath.ts]: Also can this be a separate PR? It affects all languages, not just .NET Core - and it's mostly copied so it should be easy to get approved/merged.

**Comment by EricJizbaMSFT** [comment]
Closing because this was split into smaller PRs
